python/lsst/ts/dsm/dsm_csc.py

```diff
@@ -1,34 +1,388 @@
+import asyncio
+import csv
 import logging
+import os
+import pathlib
 import tempfile
+import yaml
+
+import aionotify
 
 from lsst.ts import salobj
 
+from . import utils
+
 __all__ = ['DSMCSC']
 
 
-class DSMCSC(salobj.BaseCsc):
+class DSMCSC(salobj.ConfigurableCsc):
     """
     Commandable SAL Component to interface with the LSST DSMs.
     """
 
-    def __init__(self, index, initial_state, initial_simulation_mode=0):
+    def __init__(self, index, config_dir=None, initial_state=salobj.State.STANDBY,
+                 initial_simulation_mode=0):
         """
         Initialize DSM CSC.
+
         Parameters
         ----------
-        index : int
+        index : `int`
             Index for the DSM. This enables the control of multiple DSMs.
+        config_dir : `str`, optional
+            The location of the CSC configuration files.
+        initial_state : `lsst.ts.salobj.State`, optional
+            State to place CSC in after initialization.
+        initial_simulation_mode : `int`, optional
+            Flag to determine mode of operation.
         """
+        schema_path = pathlib.Path(__file__).resolve().parents[4].joinpath("schema", "DSM.yaml")
+
         self.telemetry_directory = None
-        self.telemetry_loop_running = False
-        self.telemetry_loop_task = None
-        super().__init__("DSM", index, initial_state=initial_state,
+        # self.telemetry_loop_running = False
+        self.telemetry_loop_task = salobj.make_done_future()
+        self.telemetry_watcher = aionotify.Watcher()
+        self.simulated_telemetry_ui_config_written = False
+        # self.simulated_telemetry_loop_running = False
+        self.simulated_telemetry_loop_task = salobj.make_done_future()
+        self.simulation_loop_time = None
+        self.config = None
+
+        # self.loop_die_timeout = 5  # how long to wait for the loops to die?
+
+        super().__init__("DSM", index, schema_path=schema_path, config_dir=config_dir,
+                         initial_state=initial_state,
                          initial_simulation_mode=initial_simulation_mode)
 
         ch = logging.StreamHandler()
         self.log.addHandler(ch)
 
+    async def begin_exitControl(self, id_data):
+        """Begin do_exitControl; called after state changes but before command
+           acknowledged.
+
+        This method will remove the telemetry directory if in simulation mode.
+
+        Parameters
+        ----------
+        id_data : `CommandIdData`
+            Command ID and data
+        """
+        if self.simulation_mode:
+            if os.path.exists(self.telemetry_directory):
+                os.removedirs(self.telemetry_directory)
```

Comment:
I suggest renaming this method to `async def close_tasks(self)` becaus that is the standard way to cleanup at shutdown. You can also cancel all background tasks if you want to be paranoid, but the code in report_summary_state probably suffices for that.

Reply:
Done.